File: src/DynamicImages/Client/src/models/relative-layout.ts
import type { DiPosition, DiRelativeReference, RelativeEdge } from "../api/types.js";
import { anchorToTopLeft, axisX, axisY, composeAnchor, type Box } from "./anchor.js";

/**
 * The client half of relative positioning. `resolvePosition` mirrors
 * Core/Rendering/RelativeLayout.cs exactly, and both are tested against the shared fixture in
 * relative-layout-fixtures.json:
 *
 * - A tracked axis becomes an edge coordinate with the anchor forced on that axis (below = top
 *   edge at the reference's bottom + gap, above = bottom edge at its top - gap, rightOf = left
 *   edge at its right + gap, leftOf = right edge at its left - gap). The other axis keeps the
 *   layer's own coordinate and anchor component.
 * - A reference that drew nothing is skipped and the chain moves on to whatever that layer
 *   tracks on the same axis, keeping this layer's own gap. Nothing left: the layer's own coordinate.
 * - A layer on a reference cycle is absolute on both axes.
 * - An edge on the wrong axis, or a reference to a missing layer, leaves the axis absolute.
 */

export type Axis = "x" | "y";

/** Just what the resolver needs of a layer, so tests need not build whole layers. */
export interface Positioned {
  key: string;
  position: DiPosition;
}

export interface Size {
  width: number;
  height: number;
}

/** A layer's position once its references are resolved, plus the box it occupies. */
export interface ResolvedLayer {
  position: DiPosition;
  box: Box;
}

export const DEFAULT_RELATIVE_GAP = 10;

export function isTracked(position: DiPosition, axis: Axis): boolean {
  return axis === "x" ? !!position.relativeX : !!position.relativeY;
}

export function isRelative(position: DiPosition): boolean {
  return !!position.relativeX || !!position.relativeY;
}

export function referenceOn(position: DiPosition, axis: Axis): DiRelativeReference | null | undefined {
  return axis === "x" ? position.relativeX : position.relativeY;
}

function isVerticalEdge(edge: RelativeEdge): boolean {
  return edge === "below" || edge === "above";
}

function references(position: DiPosition): DiRelativeReference[] {
  const result: DiRelativeReference[] = [];
  if (position.relativeX) result.push(position.relativeX);
  if (position.relativeY) result.push(position.relativeY);
  return result;
}

/** Layers by key, first occurrence winning, matching the server's index. */
export function indexLayers<T extends Positioned>(layers: readonly T[]): Map<string, T> {
  const index = new Map<string, T>();
  for (const layer of layers) {
    if (!index.has(layer.key)) index.set(layer.key, layer);
  }
  return index;
}

/** Whether following references out of this layer, on either axis, leads back to it. */
export function isOnCycle(key: string, layersByKey: ReadonlyMap<string, Positioned>): boolean {
  const start = layersByKey.get(key);
  if (!start) return false;

  const visited = new Set<string>();
  const pending = references(start.position).map((reference) => reference.layerKey);

  while (pending.length > 0) {
    const current = pending.pop()!;
    if (current === key) return true;
    if (visited.has(current)) continue;
    visited.add(current);

    const layer = layersByKey.get(current);
    if (layer) pending.push(...references(layer.position).map((reference) => reference.layerKey));
  }

  return false;
}

/**
 * The absolute position of a layer given where its references landed. `boundsOf` returns
 * undefined for a layer that drew nothing, which sends the chain on to the next reference.
 */
export function resolvePosition(
  layer: Positioned,
  layersByKey: ReadonlyMap<string, Positioned>,
  boundsOf: (key: string) => Box | undefined,
): DiPosition {
  const position = layer.position;
  if (!isRelative(position)) return position;

  // A layer on a cycle has no well-defined answer, so it keeps its own coordinates on both axes.
  if (isOnCycle(layer.key, layersByKey)) {
    return { x: position.x, y: position.y, anchor: position.anchor };
  }

  let x = position.x;
  let y = position.y;
  let factorX = axisX(position.anchor);
  let factorY = axisY(position.anchor);

  const horizontal = resolveAxis(layer, position.relativeX, false, layersByKey, boundsOf);
  if (horizontal) {
    x = horizontal.coordinate;
    factorX = horizontal.factor;
  }

  const vertical = resolveAxis(layer, position.relativeY, true, layersByKey, boundsOf);
  if (vertical) {
    y = vertical.coordinate;
    factorY = vertical.factor;
  }

  return { x, y, anchor: composeAnchor(factorX, factorY) };
}

function resolveAxis(
  layer: Positioned,
  reference: DiRelativeReference | null | undefined,
  vertical: boolean,
  layersByKey: ReadonlyMap<string, Positioned>,
  boundsOf: (key: string) => Box | undefined,
): { coordinate: number; factor: number } | undefined {
  if (!reference || isVerticalEdge(reference.edge) !== vertical) return undefined;

  const visited = new Set<string>([layer.key]);
  let current = reference.layerKey;

  while (!visited.has(current)) {
    visited.add(current);

    const target = layersByKey.get(current);
    if (!target) return undefined;

    const bounds = boundsOf(current);
    if (bounds) {
      // The gap is always this layer's own, however far up the chain the answer came from.
      switch (reference.edge) {
        case "below":
          return { coordinate: bounds.y + bounds.height + reference.gap, factor: 0 };
        case "above":
          return { coordinate: bounds.y - reference.gap, factor: 1 };
        case "rightOf":
          return { coordinate: bounds.x + bounds.width + reference.gap, factor: 0 };
        default:
          return { coordinate: bounds.x - reference.gap, factor: 1 };
      }
    }

    // The reference drew nothing: track whatever it tracks on this axis instead.
    const next = vertical ? target.position.relativeY : target.position.relativeX;
    if (!next || isVerticalEdge(next.edge) !== vertical) return undefined;
    current = next.layerKey;
  }

  return undefined;
}

/**
 * Resolves every layer at once, for the designer. A reference's box is its own resolved position
 * plus its size; `isAbsent` says which layers draw nothing (hidden, or the server measured the
 * layout and reported no bounds for them), which is what sends a tracker up the chain before
 * the server has had its say.
 */
export function resolveAll<T extends Positioned>(
  layers: readonly T[],
  sizeOf: (layer: T) => Size,
  isAbsent: (layer: T) => boolean,
): Map<string, ResolvedLayer> {
  const layersByKey = indexLayers(layers);
  const resolved = new Map<string, ResolvedLayer>();
  const inProgress = new Set<string>();

  const resolve = (layer: T): ResolvedLayer => {
    const done = resolved.get(layer.key);
    if (done) return done;

    let position: DiPosition;

    if (inProgress.has(layer.key)) {
      // Cannot happen once cycles resolve as absolute, but a guard beats a stack overflow.
      position = { x: layer.position.x, y: layer.position.y, anchor: layer.position.anchor };
    } else {
      inProgress.add(layer.key);
      position = resolvePosition(layer, layersByKey, (key) => {
        const reference = layersByKey.get(key);
        return reference && !isAbsent(reference) ? resolve(reference).box : undefined;
      });
      inProgress.delete(layer.key);
    }

    const size = sizeOf(layer);
    const topLeft = anchorToTopLeft(position, size.width, size.height);
    const entry = { position, box: { x: topLeft.x, y: topLeft.y, width: size.width, height: size.height } };

    resolved.set(layer.key, entry);
    return entry;
  };

  for (const layer of layers) resolve(layer);

  return resolved;
}

/**
 * Removes the link on one axis, baking the resolved coordinate and anchor component into the
 * position so the layer stays exactly where it was. Without a resolved position (nothing has
 * been laid out yet) the stored fallback coordinate is all there is.
 */
export function detach(position: DiPosition, axis: Axis, resolved: DiPosition | undefined): DiPosition {
  if (axis === "x") {
    return {
      ...position,
      relativeX: null,
      x: resolved ? Math.round(resolved.x) : position.x,
      anchor: resolved ? composeAnchor(axisX(resolved.anchor), axisY(position.anchor)) : position.anchor,
    };
  }

  return {
    ...position,
    relativeY: null,
    y: resolved ? Math.round(resolved.y) : position.y,
    anchor: resolved ? composeAnchor(axisX(position.anchor), axisY(resolved.anchor)) : position.anchor,
  };
}

File: src/DynamicImages/Core/Rendering/RelativeLayout.cs
using Umbraco.Community.DynamicImages.Core.Models;
using Umbraco.Community.DynamicImages.Core.Models.Layers;
using Umbraco.Community.DynamicImages.Core.Services;

namespace Umbraco.Community.DynamicImages.Core.Rendering;

/// <summary>
/// Resolves positions that track another layer (<see cref="Position.RelativeX"/> and
/// <see cref="Position.RelativeY"/>) into absolute ones.
/// <para>
/// The rules, mirrored exactly by <c>Client/src/models/relative-layout.ts</c> and tested on both
/// sides against <c>relative-layout-fixtures.json</c>:
/// </para>
/// <list type="bullet">
/// <item>A tracked axis becomes an edge coordinate with the anchor forced on that axis: below =
/// top edge at the reference's bottom plus the gap, above = bottom edge at its top minus the gap,
/// rightOf = left edge at its right plus the gap, leftOf = right edge at its left minus the gap.
/// The other axis keeps the layer's own coordinate and anchor component.</item>
/// <item>A reference that drew nothing (empty value, hidden, visibility rule) is skipped and the
/// chain moves on to whatever that layer tracks on the same axis, keeping this layer's own gap.
/// With nothing left, the axis uses the layer's own coordinate.</item>
/// <item>A layer on a reference cycle resolves as absolute on both axes, so the result never
/// depends on z-order or evaluation order.</item>
/// <item>An edge that belongs to the other axis, or a reference to a layer that no longer exists,
/// leaves the axis absolute; both are reported by <see cref="Problems"/>.</item>
/// </list>
/// Pure and static, so it needs no Umbraco services to test.
/// </summary>
public static class RelativeLayout
{
    /// <summary>Whether any layer in the template tracks another, which is what turns the measure pass on.</summary>
    public static bool IsUsed(Template template)
        => template.Layers.Any(layer => layer.Position.IsRelative);

    /// <summary>Layers by key, first occurrence winning should a document ever carry a duplicate.</summary>
    public static IReadOnlyDictionary<Guid, LayerBase> Index(Template template)
    {
        var index = new Dictionary<Guid, LayerBase>(template.Layers.Count);
        foreach (var layer in template.Layers) index.TryAdd(layer.Key, layer);
        return index;
    }

    /// <summary>
    /// Every layer some other layer references, directly or through a chain, ordered so that a
    /// layer comes after everything it references. That is the order the renderer measures them
    /// in before drawing, so a reference can sit anywhere in z-order. Layers on a cycle are still
    /// included - they resolve as absolute, so their order among themselves does not matter.
    /// </summary>
    public static IReadOnlyList<LayerBase> MeasureOrder(Template template)
    {
        var layersByKey = Index(template);
        var referenced = new HashSet<Guid>();

        foreach (var layer in template.Layers)
        {
            foreach (var reference in References(layer)) referenced.Add(reference.LayerKey);
        }

        var order = new List<LayerBase>();
        var state = new Dictionary<Guid, bool>(); // false = visiting, true = done

        void Visit(Guid key)
        {
            if (state.ContainsKey(key)) return; // done, or a back-edge inside a cycle
            if (!layersByKey.TryGetValue(key, out var layer)) return;

            state[key] = false;
            foreach (var reference in References(layer)) Visit(reference.LayerKey);
            state[key] = true;
            order.Add(layer);
        }

        // Walk in z-order so the result is stable for a given document.
        foreach (var layer in template.Layers)
        {
            if (referenced.Contains(layer.Key)) Visit(layer.Key);
        }

        return order;
    }

    /// <summary>
    /// The absolute position of <paramref name="layer"/> given where its references landed.
    /// <paramref name="boundsOf"/> returns null for a layer that drew nothing (or has not been
    /// measured), which is what sends the chain on to the next reference.
    /// </summary>
    public static Position Resolve(
        LayerBase layer,
        IReadOnlyDictionary<Guid, LayerBase> layersByKey,
        Func<Guid, LayerBounds?> boundsOf)
    {
        var position = layer.Position;
        if (!position.IsRelative) return position;

        // A layer on a cycle has no well-defined answer, so it keeps its own coordinates - on
        // both axes, so the outcome cannot depend on which member happened to draw first.
        if (IsOnCycle(layer.Key, layersByKey))
        {
            return new Position { X = position.X, Y = position.Y, Anchor = position.Anchor };
        }

        var x = position.X;
        var y = position.Y;
        var axisX = AnchorMath.AxisX(position.Anchor);
        var axisY = AnchorMath.AxisY(position.Anchor);

        var horizontal = ResolveAxis(layer, position.RelativeX, isVertical: false, layersByKey, boundsOf);
        if (horizontal is { } h)
        {
            x = h.Coordinate;
            axisX = h.Factor;
        }

        var vertical = ResolveAxis(layer, position.RelativeY, isVertical: true, layersByKey, boundsOf);
        if (vertical is { } v)
        {
            y = v.Coordinate;
            axisY = v.Factor;
        }

        return new Position { X = x, Y = y, Anchor = AnchorMath.Compose(axisX, axisY) };
    }

    private static (float Coordinate, float Factor)? ResolveAxis(
        LayerBase layer,
        RelativeReference? reference,
        bool isVertical,
        IReadOnlyDictionary<Guid, LayerBase> layersByKey,
        Func<Guid, LayerBounds?> boundsOf)
    {
        if (reference is null || IsVerticalEdge(reference.Edge) != isVertical) return null;

        var visited = new HashSet<Guid> { layer.Key };
        var current = reference.LayerKey;

        while (visited.Add(current))
        {
            if (!layersByKey.TryGetValue(current, out var target)) return null;

            var bounds = boundsOf(current);
            if (bounds is not null)
            {
                // What the reference covered on the canvas: its rotated footprint, or the box
                // itself when it is not rotated. The gap is always this layer's own, however far
                // up the chain the answer came from.
                var extent = bounds.Extent();

                return reference.Edge switch
                {
                    RelativeEdge.Below => (extent.Y + extent.Height + reference.Gap, 0f),
                    RelativeEdge.Above => (extent.Y - reference.Gap, 1f),
                    RelativeEdge.RightOf => (extent.X + extent.Width + reference.Gap, 0f),
                    _ => (extent.X - reference.Gap, 1f)
                };
            }

            // The reference drew nothing: track whatever it tracks on this axis instead.
            var next = isVertical ? target.Position.RelativeY : target.Position.RelativeX;
            if (next is null || IsVerticalEdge(next.Edge) != isVertical) return null;
            current = next.LayerKey;
        }

        return null;
    }

    /// <summary>
    /// Whether following references out of <paramref name="key"/> (on either axis) leads back to
    /// it. Small templates make the direct walk cheaper than anything cleverer, and it is easy to
    /// mirror on the client.
    /// </summary>
    public static bool IsOnCycle(Guid key, IReadOnlyDictionary<Guid, LayerBase> layersByKey)
    {
        if (!layersByKey.TryGetValue(key, out var start)) return false;

        var visited = new HashSet<Guid>();
        var pending = new Stack<Guid>(References(start).Select(r => r.LayerKey));

        while (pending.Count > 0)
        {
            var current = pending.Pop();
            if (current == key) return true;
            if (!visited.Add(current) || !layersByKey.TryGetValue(current, out var layer)) continue;

            foreach (var reference in References(layer)) pending.Push(reference.LayerKey);
        }

        return false;
    }

    /// <summary>Validation warnings for references that cannot do what they say.</summary>
    public static IEnumerable<ValidationIssue> Problems(Template template)
    {
        var layersByKey = Index(template);

        foreach (var layer in template.Layers)
        {
            var name = string.IsNullOrWhiteSpace(layer.Name) ? layer.TypeAlias : layer.Name;
            var selfReference = false;

            foreach (var (reference, isVertical) in new[] { (layer.Position.RelativeX, false), (layer.Position.RelativeY, true) })
            {
                if (reference is null) continue;

                var axis = isVertical ? "Y" : "X";

                if (reference.LayerKey == layer.Key)
                {
                    selfReference = true;
                    yield return new ValidationIssue(ValidationSeverity.Warning, "RelativeSelf",
                        $"Layer '{name}' is positioned relative to itself on {axis}, so that axis stays where it is.", layer.Key);
                }
                else if (!layersByKey.ContainsKey(reference.LayerKey))
                {
                    yield return new ValidationIssue(ValidationSeverity.Warning, "RelativeMissing",
                        $"Layer '{name}' is positioned relative to a layer that no longer exists on {axis}, so that axis stays where it is.", layer.Key);
                }

                if (IsVerticalEdge(reference.Edge) != isVertical)
                {
                    yield return new ValidationIssue(ValidationSeverity.Warning, "RelativeEdgeMismatch",
                        $"Layer '{name}' uses the '{EdgeName(reference.Edge)}' edge on {axis}, which is the other axis, so that axis stays where it is.", layer.Key);
                }
            }

            if (!selfReference && layer.Position.IsRelative && IsOnCycle(layer.Key, layersByKey))
            {
                yield return new ValidationIssue(ValidationSeverity.Warning, "RelativeCycle",
                    $"Layer '{name}' is part of a loop of relative positions, so it is placed at its own coordinates.", layer.Key);
            }
        }
    }

    private static IEnumerable<RelativeReference> References(LayerBase layer)
    {
        if (layer.Position.RelativeX is { } x) yield return x;
        if (layer.Position.RelativeY is { } y) yield return y;
    }

    private static bool IsVerticalEdge(RelativeEdge edge) => edge is RelativeEdge.Below or RelativeEdge.Above;

    private static string EdgeName(RelativeEdge edge) => edge switch
    {
        RelativeEdge.Below => "below",
        RelativeEdge.Above => "above",
        RelativeEdge.RightOf => "rightOf",
        _ => "leftOf"
    };
}
